File: src/LoneWolf/Choices.hs
{-# LANGUAGE TupleSections #-}

module LoneWolf.Choices where

import Control.Lens
import Data.List
import qualified Data.Map.Strict as M
import Data.Maybe (fromMaybe)
import qualified LoneWolf.Cartwheel
import LoneWolf.Chapter
import LoneWolf.Character
import LoneWolf.Combat (getRatio)
import LoneWolf.CombatChart (hits)
import LoneWolf.Rules (check, getMaxHp, updateSimple)
import Solver (regroup)

importantItem :: Item -> CharacterConstant -> CharacterVariable -> Bool
importantItem i cconstant cvariable =
  case i of
    Weapon Sommerswerd -> True
    Weapon MagicSpear -> True
    -- there is a magic mace in book 05, but due the way the chapters are made we can't decide in advance if this mace is the magic mace or not :(
    Weapon Mace | _bookid cconstant == Book05 -> True
    Weapon w -> case usedWeapon cconstant cvariable of
      WithSkill _ -> False
      NoWeapon -> True
      WithoutSkill _ -> WeaponSkill w `elem` cconstant ^. discipline
    _ -> True

-- some items will force branches (particularly in book05, with the black crystal cube, and should not always be picked up)
potentiallybad :: Item -> Bool
potentiallybad i = i == blackCubeB05

data CanTake
  = SpaceAvailable Int
  | Mustdrop [Item]
  | Nope
  deriving (Show, Eq)

canTake :: Item -> CharacterConstant -> CharacterVariable -> CanTake
canTake item cconstant cvariable =
  case itemSlot item of
    BackpackSlot
      | itemsInBackpack > 8 -> error ("canTake: too many items in backpack -> " ++ show cvariable)
      | itemsInBackpack == 8 -> Mustdrop (map fst backpackItems)
      | otherwise -> SpaceAvailable (8 - itemsInBackpack)
    WeaponSlot -> case weapons of
      [] -> SpaceAvailable 2
      [_] -> SpaceAvailable 1
      lst -> Mustdrop $
        map Weapon $ case usedWeapon cconstant cvariable of
          WithSkill w -> filter (/= w) lst
          WithoutSkill Sommerswerd -> filter (/= Sommerswerd) lst
          _ -> lst
    PouchSlot -> SpaceAvailable 50
    SpecialSlot ->
      if hasItem item inventory
        then Nope
        else SpaceAvailable 1
  where
    inventory = cvariable ^. equipment
    weapons = getWeapons inventory
    backpackItems = filter ((== BackpackSlot) . itemSlot . fst) (items inventory)
    itemsInBackpack = sum (map snd backpackItems)

flattenDecision :: CharacterConstant -> CharacterVariable -> Decision -> [([String], ChapterOutcome)]
flattenDecision cconstant cvariable d =
  let inventory = cvariable ^. equipment
      takeItem item = withEffects [GainItem item 1]
      withEffects effects nxt = do
        let nvariable = foldl' (updateSimple cconstant) cvariable effects
        (dsc, o) <- flattenDecision cconstant nvariable nxt
        return (show effects : dsc, Simple effects o)
   in case d of
        _ -- can be healed, finally!
          | hasFlag Poisonned2 cvariable && hasItem Laumspur inventory ->
            let nvariable = updateSimple cconstant cvariable (HealPlayer 4) & equipment %~ delItem Laumspur 1 & flag Poisonned2 .~ False
             in flattenDecision cconstant nvariable d
        NoDecision o
          | Fight (FightDetails _ _ _ mds) _ <- o, NoPotion `elem` mds -> [([], o)]
          | hasCombat o && hasItem StrengthPotion4 inventory && hasItem StrengthPotion inventory ->
            [ (["don't use strength potion"], o),
              (["use strength potion"], Simple [LoseItem StrengthPotion4 1, SetFlag PotentStrengthPotionActive] o),
              (["use potent strength potion"], Simple [LoseItem StrengthPotion4 1, SetFlag PotentStrengthPotionActive] o)
            ]
          | hasCombat o && hasItem StrengthPotion inventory ->
            [ (["don't use strength potion"], o),
              (["use strength potion"], Simple [LoseItem StrengthPotion 1, SetFlag StrengthPotionActive] o)
            ]
          | hasCombat o && hasItem StrengthPotion4 inventory ->
            [ (["don't use strength potion"], o),
              (["use potent strength potion"], Simple [LoseItem StrengthPotion4 1, SetFlag PotentStrengthPotionActive] o)
            ]
          | otherwise -> [([], o)]
        RetrieveEquipment nxt ->
          -- there should be less items in the current inventory than in the stored, that's why it is done that way
          let CharacterVariable e flgs cure stoe = cvariable
              itms = items cure
           in flattenDecision cconstant (CharacterVariable e flgs stoe 0) (foldr (uncurry CanTake) nxt itms)
        AfterCombat nxt
          | hasItem Potion6Hp inventory && missingHp >= 6 ->
            nopotion ++ withEffects [HealPlayer 6, LoseItem Potion6Hp 1] d
          | hasItem Potion5Hp inventory && missingHp >= 5 ->
            nopotion ++ withEffects [HealPlayer 5, LoseItem Potion5Hp 1] d
          | not book01 && hasItem Laumspur inventory && missingHp >= 4 ->
            nopotion ++ withEffects [HealPlayer 4, LoseItem Laumspur 1] d
          | hasItem Potion4Hp inventory && missingHp >= 4 ->
            nopotion ++ withEffects [HealPlayer 4, LoseItem Potion4Hp 1] d
          | hasItem Potion2Hp inventory && missingHp >= 2 ->
            nopotion ++ withEffects [HealPlayer 2, LoseItem Potion2Hp 1] d
          | _bookid cconstant == Book05 && hasItem oedeHerb 1 && missingHp >= 10 ->
            nopotion ++ withEffects [HealPlayer 10, LoseItem oedeHerb 1] d
          | otherwise -> nopotion
          where
            nopotion = flattenDecision cconstant cvariable nxt
            missingHp = getMaxHp cconstant cvariable - cvariable ^. curendurance
            book01 = _bookid cconstant == Book01
        Conditional bc d' ->
          if check cconstant cvariable bc
            then flattenDecision cconstant cvariable d'
            else []
        Special Portholes -> [([], Goto 197)]
        Special Cartwheel -> cartwheel (cvariable ^. equipment . gold)
        EvadeFight nrounds cid fdetails co ->
          [ (["no evasion"], Fight fdetails co),
            (["evasion"], Fight (fdetails & fightMod %~ ((if nrounds > 0 then Timed nrounds (Evaded cid) else Evaded cid) :)) co)
          ]
        Decisions lst -> do
          (cdesc, d') <- lst
          (alldesc, o) <- flattenDecision cconstant cvariable d'
          return (cdesc : alldesc, o)
        CanTake item count nxt
          | item == Gold -> withEffects [GainItem Gold count] nxt
          | count == 0 -> flattenDecision cconstant cvariable nxt
          | count == 1 ->
            case canTake item cconstant cvariable of
              Nope -> notake
              SpaceAvailable n | n <= 0 -> error "space available invariant wrong"
              SpaceAvailable _ -> takeItem item nxt'
              Mustdrop drops
                | not (importantItem item cconstant cvariable) -> notake
                | otherwise -> do
                  itemDropped <- drops
                  withEffects [LoseItem itemDropped 1, GainItem item 1] nxt'
          | otherwise -> flattenDecision cconstant cvariable (CanTake item 1 (CanTake item (count - 1) nxt))
          where
            notake = flattenDecision cconstant cvariable nxt
            nxt' = CanTake item (count - 1) nxt
        Canbuy item price nxt
          | cvariable ^. equipment . gold < price ->
            nobuy
          | not (importantItem item cconstant cvariable) ->
            nobuy
          | otherwise ->
            case canTake item cconstant cvariable of
              Nope -> nobuy
              SpaceAvailable _ -> withEffects [LoseItem Gold price, GainItem item 1] d
              Mustdrop drops -> do
                itemDropped <- drops
                withEffects [LoseItem Gold price, LoseItem itemDropped 1, GainItem item 1] d
          where
            nobuy = flattenDecision cconstant cvariable nxt
        Cansell item price nxt
          | not (hasItem item inventory) ->
            flattenDecision cconstant cvariable nxt
          | otherwise -> nosell ++ sell
          where
            nosell = flattenDecision cconstant cvariable nxt
            sell = withEffects [GainItem Gold price, LoseItem item 1] nxt
        Special B05S127 ->
          let cs = fromIntegral (cconstant ^. combatSkill)
              borne x
                | x > 1 = 1
                | x < 0 = 0
                | otherwise = x
              b1 = borne ((cs - 10) / 10)
              b2 = borne ((20 - cs) / 10)
           in if b1 + b2 /= 1
                then error "SPECIAL B05S127"
                else [(["No decision"], Randomly [(b1, Goto 159), (b2, Goto 93)])]
        RemoveItemFrom BackpackSlot n nxt
          | n <= 0 -> flattenDecision cconstant cvariable nxt
          | otherwise ->
            let allbackpackitems = filter ((== BackpackSlot) . itemSlot . fst) (items inventory)
             in if sum (map snd allbackpackitems) < n
                  then []
                  else do
                    (todrop, _) <- allbackpackitems
                    let nvariable = updateSimple cconstant cvariable (LoseItem todrop 1)
                    (dsc, o) <- flattenDecision cconstant nvariable (RemoveItemFrom BackpackSlot (n - 1) nxt)
                    return (("drop " ++ show todrop) : dsc, Simple [LoseItem todrop 1] o)
        LoseItemFrom BackpackSlot n nxt ->
          let allbackpackitems = filter ((== BackpackSlot) . itemSlot . fst) (items inventory)
           in if n <= 0 || null allbackpackitems
                then flattenDecision cconstant cvariable nxt
                else do
                  (todrop, _) <- allbackpackitems
                  let nvariable = updateSimple cconstant cvariable (LoseItem todrop 1)
                  (dsc, o) <- flattenDecision cconstant nvariable (LoseItemFrom BackpackSlot (n - 1) nxt)
                  return (("lost " ++ show todrop) : dsc, Simple [LoseItem todrop 1] o)
        Special B05S357 ->
          let lwe = cvariable ^. curendurance
              ratio = getRatio cconstant cvariable (FightDetails "Platform Sentry" 15 23 [CombatBonus (-2)])
              result n = Randomly $ do
                ((lwe', mcid), p) <- bk05c357map M.! (ratio, lwe, 23)
                let outcome c = Simple [DamagePlayer (lwe - lwe')] (Goto c)
                pure (p, outcome (fromMaybe n mcid))
           in [(["Search the sentry body"], result 207), (["Ignore the sentry body"], result 224)]
        x -> error ("TODO: " ++ show x)

bk05c357map :: M.Map (CombatSkill, Endurance, Endurance) [((Endurance, Maybe ChapterId), Rational)]
bk05c357map = M.fromList [((cs, elw, een), bk05c357 cs elw een) | cs <- [-10 .. 40], elw <- [0 .. 40], een <- [0 .. 24]]

bk05c357 :: CombatSkill -> Endurance -> Endurance -> [((Endurance, Maybe ChapterId), Rational)]
bk05c357 ratio lwhp ophp
  | lwhp <= 0 = [((0, Nothing), 1)]
  | ophp <= 0 = [((lwhp, Nothing), 1)]
  | otherwise = ((lwhp, Just 293), 1 / 10) : regroup nxt
  where
    curp = 9 / 100
    nxt = do
      (dmgopponent, dmglw) <- hits ratio
      let lwhp' = lwhp - dmglw
          ophp' = ophp - dmgopponent
      if lwhp' <= 0
        then [((0, Nothing), curp)]
        else
          if ophp' <= 0
            then [((lwhp', Nothing), curp)]
            else do
              (o, p) <- case M.lookup (ratio, lwhp', ophp') bk05c357map of
                Just x -> x
                Nothing -> error ("Could not find key " ++ show (ratio, lwhp', ophp'))
              pure (o, p * curp)

cartwheel :: Price -> [([String], ChapterOutcome)]
cartwheel m
  | m >= 22 = [([], Goto 136)]
  | otherwise =
    let cmoney = m + 1 -- free token!
        target = 22
        cdesc = "target: " ++ show target
        res = do
          (newmoney, proba) <- LoneWolf.Cartwheel.solveFor target cmoney
          return $
            (proba,) $
              if newmoney == 0
                then GameLost
                else Simple [GainItem Gold (newmoney - m)] (Goto 136)
     in [([cdesc], Randomly res)]

File: gamebooksolver-explorer/src/lonewolf/choices.rs
use crate::lonewolf::mini::{max_hp, Discipline, Flag, Slot, Weapon};
use crate::lonewolf::rules::update_simple;
use crate::{
  Book, ChapterOutcome, CharacterConstant, CharacterVariable, Decision, Endurance, FightDetails,
  FightModifier, Item, SimpleOutcome,
};

enum UsedWeapon {
  WithSkill(Weapon),
  WithoutSkill(Weapon),
  NoWeapon,
}

fn used_weapon(ccst: &CharacterConstant, cvar: &CharacterVariable) -> UsedWeapon {
  let wskill = ccst
    .discipline
    .iter()
    .filter_map(|d| match d {
      Discipline::WeaponSkill(w) => Some(*w),
      _ => None,
    })
    .next();
  if cvar
    .cequipment
    .has_itemb(&Item::Weapon(Weapon::Sommerswerd))
  {
    if wskill == Some(Weapon::Sword)
      || wskill == Some(Weapon::ShortSword)
      || wskill == Some(Weapon::BroadSword)
    {
      UsedWeapon::WithSkill(Weapon::Sommerswerd)
    } else {
      UsedWeapon::WithoutSkill(Weapon::Sommerswerd)
    }
  } else {
    let weapons = cvar.cequipment.weapons();
    if weapons.is_empty() {
      UsedWeapon::NoWeapon
    } else if let Some(w) = weapons.iter().filter(|cw| Some(**cw) == wskill).next() {
      UsedWeapon::WithSkill(*w)
    } else {
      UsedWeapon::WithoutSkill(weapons[0])
    }
  }
}

fn important_item(i: &Item, ccst: &CharacterConstant, cvar: &CharacterVariable) -> bool {
  match i {
    Item::Weapon(Weapon::Sommerswerd) => true,
    Item::Weapon(Weapon::MagicSpear) => true,
    Item::Weapon(w) => {
      if w == &Weapon::Mace && ccst.bookid == Book::Book05 {
        true
      } else {
        match used_weapon(ccst, cvar) {
          UsedWeapon::WithSkill(_) => false,
          UsedWeapon::NoWeapon => true,
          UsedWeapon::WithoutSkill(_) => ccst.discipline.contains(&Discipline::WeaponSkill(*w)),
        }
      }
    }
    _ => true,
  }
}

enum CanTake {
  SpaceAvailable(u8),
  MustDrop(Vec<Item>),
  Nope,
}

fn can_take(item: &Item, ccst: &CharacterConstant, cvar: &CharacterVariable) -> CanTake {
  match item.slot() {
    Slot::Special => {
      if cvar.cequipment.has_itemb(item) {
        CanTake::Nope
      } else {
        CanTake::SpaceAvailable(1)
      }
    }
    Slot::Pouch => CanTake::SpaceAvailable(50),
    Slot::Weapon => {
      let weapons = cvar.cequipment.weapons();
      match weapons.len() {
        0 => CanTake::SpaceAvailable(2),
        1 => CanTake::SpaceAvailable(1),
        _ => CanTake::MustDrop(match used_weapon(ccst, cvar) {
          UsedWeapon::WithSkill(w) => weapons
            .into_iter()
            .filter_map(|x| if x == w { None } else { Some(Item::Weapon(x)) })
            .collect(),
          _ => weapons
            .into_iter()
            .filter(|x| *x != Weapon::Sommerswerd)
            .map(Item::Weapon)
            .collect(),
        }),
      }
    }
    Slot::Backpack => {
      let mut bpitems = cvar.cequipment.in_backpack();
      if bpitems.len() > 8 {
        panic!("too many items in backpack :(")
      }
      if bpitems.len() < 8 {
        CanTake::SpaceAvailable(8 - bpitems.len() as u8)
      } else {
        bpitems.dedup();
        CanTake::MustDrop(bpitems)
      }
    }
  }
}

fn has_combat(o: &ChapterOutcome) -> Option<&FightDetails> {
  use ChapterOutcome::*;
  match o {
    Fight(fd, _) => Some(fd),
    OneRound(fd, _, _, _) => Some(fd),
    Randomly(choices) => choices.iter().map(|x| &x.1).flat_map(has_combat).next(),
    Conditionally(conds) => conds.iter().map(|x| &x.1).flat_map(has_combat).next(),
    Simple(_, o2) => has_combat(o2),
    Goto(_) => None,
  }
}

pub fn flatten_decision(
  ccst: &CharacterConstant,
  cvar: &CharacterVariable,
  dec: &Decision,
) -> Vec<(Vec<String>, ChapterOutcome)> {
  // can be healed!
  if cvar.flags.has(Flag::Poisonned2) && cvar.cequipment.has_itemb(&Item::Laumspur) {
    let mut nv: CharacterVariable = cvar.clone();
    nv.flags.unset(Flag::Poisonned2);
    nv.cequipment.del_item(&Item::Laumspur, 1);
    return flatten_decision(ccst, &nv, dec);
  }
  let with_effect = |sos: &[SimpleOutcome], nxt: &Decision| -> Vec<(Vec<String>, ChapterOutcome)> {
    let mut nv = cvar.clone();
    for so in sos {
      update_simple(&mut nv, ccst, so);
    }
    flatten_decision(ccst, &nv, nxt).into_iter().map(|(mut dsc, out)| {
      dsc.insert(0, format!("{:?}", sos));
      (dsc, out)
    }).collect()
  };
  match dec {
    Decision::None(o) => {
      if let Some(fd) = has_combat(o) {
        if fd.fight_mod.contains(&FightModifier::NoPotion) {
          vec![(Vec::new(), o.clone())]
        } else {
          let mut out = vec![(vec!["No strength potion".to_string()], o.clone())];
          if cvar.cequipment.has_itemb(&Item::StrengthPotion4) {
            out.push((
              vec!["Use potent strength potion".to_string()],
              ChapterOutcome::Simple(
                vec![
                  SimpleOutcome::SetFlag(Flag::PotentStrengthPotionActive),
                  SimpleOutcome::LoseItem(Item::StrengthPotion4, 1),
                ],
                Box::new(o.clone()),
              ),
            ))
          }
          if cvar.cequipment.has_itemb(&Item::StrengthPotion) {
            out.push((
              vec!["Use strength potion".to_string()],
              ChapterOutcome::Simple(
                vec![
                  SimpleOutcome::SetFlag(Flag::StrengthPotionActive),
                  SimpleOutcome::LoseItem(Item::StrengthPotion, 1),
                ],
                Box::new(o.clone()),
              ),
            ))
          }
          out
        }
      } else {
        vec![(Vec::new(), o.clone())]
      }
    }
    Decision::RetrieveEquipment(nxt) => {
      let mut nv: CharacterVariable = cvar.clone();
      let cur = nv.cequipment;
      nv.cequipment = nv.cprevequipment;
      let rdec: &Decision = nxt;
      let ndec: Decision = cur.items().into_iter().fold(rdec.clone(), |acc, (i, q)| {
        Decision::CanTake(i, q, Box::new(acc))
      });
      flatten_decision(ccst, &nv, &ndec)
    }
    Decision::AfterCombat(nxt) => {
      let missing_hp = max_hp(ccst, cvar) - cvar.curendurance;
      let mut out = flatten_decision(ccst, cvar, nxt);
      if missing_hp >= 6 && cvar.cequipment.has_itemb(&Item::Potion6Hp) {
        out.extend(with_effect(&[
          SimpleOutcome::HealPlayer(Endurance(6)),
          SimpleOutcome::LoseItem(Item::Potion6Hp, 1),
        ], dec));
        return out;
      }
      if missing_hp >= 5 && cvar.cequipment.has_itemb(&Item::Potion5Hp) {
        out.extend(with_effect(&[
          SimpleOutcome::HealPlayer(Endurance(5)),
          SimpleOutcome::LoseItem(Item::Potion5Hp, 1),
        ], dec));
        return out;
      }
      if missing_hp >= 4 && cvar.cequipment.has_itemb(&Item::Potion4Hp) {
        out.extend(with_effect(&[
          SimpleOutcome::HealPlayer(Endurance(4)),
          SimpleOutcome::LoseItem(Item::Potion4Hp, 1),
        ], dec));
        return out;
      }
      if ccst.bookid != Book::Book01
        && missing_hp >= 4
        && cvar.cequipment.has_itemb(&Item::Laumspur)
      {
        out.extend(with_effect(&[
          SimpleOutcome::HealPlayer(Endurance(4)),
          SimpleOutcome::LoseItem(Item::Laumspur, 1),
        ], dec));
        return out;
      }
      if missing_hp >= 2 && cvar.cequipment.has_itemb(&Item::Potion2Hp) {
        out.extend(with_effect(&[
          SimpleOutcome::HealPlayer(Endurance(2)),
          SimpleOutcome::LoseItem(Item::Potion6Hp, 1),
        ], dec));
        return out;
      }
      if ccst.bookid == Book::Book05
        && missing_hp >= 10
        && cvar.cequipment.has_itemb(&Item::GenBackpack(2))
      {
        out.extend(with_effect(&[
          SimpleOutcome::HealPlayer(Endurance(10)),
          SimpleOutcome::LoseItem(Item::GenBackpack(2), 1),
        ], dec));
        return out;
      }
      return out;
    }
  }
}
